src/app/tests/TestValveConfigurationAndControl.cpp

```diff
@@ -0,0 +1,2250 @@
+    Percent target                   = 0;
+};
+
+class TestDelegateNoLevel : public NonLevelControlDelegate
+{
+public:
+    TestDelegateNoLevel() {}
+    CHIP_ERROR HandleOpenValve(ValveStateEnum & currentState, BitMask<ValveFaultBitmap> & valveFault) override
+    {
+        ++numHandleOpenValveCalls;
+        if (simulateAsync)
+        {
+            currentState = ValveStateEnum::kTransitioning;
+        }
+        else
+        {
+            currentState = ValveStateEnum::kOpen;
+        }
+        state  = currentState;
+        target = ValveStateEnum::kOpen;
+        if (simulateOpenFailure || simulateValveFaultNoFailure)
+        {
+            valveFault = simulatedFailureBitMask;
+        }
+        if (simulateOpenFailure)
+        {
+            return CHIP_ERROR_INTERNAL;
+        }
+        return CHIP_NO_ERROR;
+    }
+    CHIP_ERROR HandleCloseValve(ValveStateEnum & currentState, BitMask<ValveFaultBitmap> & valveFault) override
+    {
+        ++numHandleCloseValveCalls;
+        if (simulateAsync)
+        {
+            currentState = ValveStateEnum::kTransitioning;
+        }
+        else
+        {
+            currentState = ValveStateEnum::kClosed;
+        }
+        state  = currentState;
+        target = ValveStateEnum::kClosed;
+        if (simulateCloseFailure || simulateValveFaultNoFailure)
+        {
+            valveFault = simulatedFailureBitMask;
+        }
+        if (simulateCloseFailure)
+        {
+            return CHIP_ERROR_INTERNAL;
+        }
+        return CHIP_NO_ERROR;
+    }
+    ValveStateEnum GetCurrentValveState() override
+    {
+        // Might want to have called more than one time before hitting the target.
+        state = target;
+        return state;
+    }
+
+    BitMask<ValveFaultBitmap> simulatedFailureBitMask =
+        BitMask<ValveFaultBitmap>(to_underlying(ValveFaultBitmap::kBlocked) | to_underlying(ValveFaultBitmap::kGeneralFault));
+    bool simulateOpenFailure         = false;
+    bool simulateCloseFailure        = false;
+    bool simulateValveFaultNoFailure = false;
+    bool simulateAsync               = false;
+    int numHandleOpenValveCalls      = 0;
+    int numHandleCloseValveCalls     = 0;
+    ValveStateEnum state             = ValveStateEnum::kUnknownEnumValue;
+    ValveStateEnum target            = ValveStateEnum::kUnknownEnumValue;
+};
+
+class MockedMatterContext : public MatterContext
+{
+public:
+    MockedMatterContext(EndpointId endpoint, PersistentStorageDelegate & persistentStorageDelegate) :
+        MatterContext(endpoint, persistentStorageDelegate)
+    {}
+    void MarkDirty(AttributeId id) override { mDirtyMarkedList.push_back(id); }
+    std::vector<AttributeId> GetDirtyList() { return mDirtyMarkedList; }
+    void ClearDirtyList() { mDirtyMarkedList.clear(); }
+    ~MockedMatterContext() { gSystemLayerAndClock.Clear(); }
+
+private:
+    // Won't handle double-marking an attribute, so don't do that in tests
```

Comment:
nit, not sure I follow this comment. This does supposed double-marking an attribute you would see it come up twice in this vector